### PR TITLE
Handle HTTP errors on non-event requests (fixes #1120)

### LINKED ISSUE #1120: No Warning That Cookies Permission Are Required
Hi,
```
debian:.config$ /opt/syncthing/syncthing --version
syncthing v0.10.12 (go1.4 linux-amd64 default) unknown-user@syncthing-builder 2014-12-16 22:33:56 UTC
```
With this and earlier versions, the UI requires cookies.  Without cookies, you can edit sync folders from the UI but the results are not saved.  When you try to restart _syncthing_ from the UI it seems to hang.

This is very confusing for the first time user.

Not that I'm bothered, but aren't there, in some jurisdictions, an obligation to warn users about cookies ?

### PR DESCRIPTION
**Pull request by calmh**
This throws up a big red box when an HTTP request fails with an error code, instead of silently ignoring it like we do know. In particular, this handles the case where the CSRF token is wrong and the user tries to make a config change. Currently that looks like it works (but the POST is rejected), with this change it shows the error and prompts the user to reload the UI.

### CODE REVIEW

File: gui/index.html [hunk 1]
```diff
@@ -416,6 +416,14 @@ <h3 class="panel-title">
     </p>
   </modal>
 
+  <!-- HTTP error modal -->
+
+  <modal id="httpError" status="danger" icon="exclamation-sign" title="{{'Connection Error' | translate}}">
+    <p translate>
+      Syncthing seems to be experiencing a problem processing your request. Please reload your browser or restart Syncthing if the problem persists.
```
Comment (AudriusButkevicius) [inline gui/index.html]: reload your browser?
Perhaps refresh the page?

File: gui/scripts/syncthing/core/controllers/syncthingController.js [hunk 2]
```diff
@@ -106,6 +105,27 @@ angular.module('syncthing.core')
...
+
+            console.log('HTTPError', arg);
+            online = false;
+            if (!restarting) {
+                if (arg.status === 0) {
+                    // A network error, not an HTTP error
+                    $scope.$emit('UIOffline');
+                } else if (arg.status >= 400 && arg.status <= 599) {
+                    // A genuine HTTP error
+                    $('#httpError').modal();
```
Comment (AudriusButkevicius) [inline gui/scripts/syncthing/core/controllers/syncthingController.js]: Shouldn't we close existing modals?

File: gui/scripts/syncthing/core/controllers/syncthingController.js [hunk 3]
```diff
@@ -106,6 +105,27 @@ angular.module('syncthing.core')
...
+        $scope.$on('HTTPError', function (event, arg) {
+            // Emitted when a HTTP call fails. We use the status code to try
+            // to figure out what's wrong.
+
+            if (navigatingAway || !online) {
+                return;
+            }
+
+            console.log('HTTPError', arg);
+            online = false;
```
Comment (AudriusButkevicius) [inline gui/scripts/syncthing/core/controllers/syncthingController.js]: Why is this necessary?

File: gui/index.html (hunk 1 above)
Comment (calmh) [inline gui/index.html]: Yeah, "refresh the page" is probably better.

File: gui/scripts/syncthing/core/controllers/syncthingController.js (hunk 2 above)
Comment (calmh) [inline gui/scripts/syncthing/core/controllers/syncthingController.js]: The stacking does work... I can close the others beforehand though, doesn't hurt.

File: gui/scripts/syncthing/core/controllers/syncthingController.js (hunk 3 above)
Comment (calmh) [inline gui/scripts/syncthing/core/controllers/syncthingController.js]: Don't want to throw up the problem dialog while we are restarting or if it's already up? I didn't think too much about it, just copied the logic from the network problem dialog which is triggered under similar circumstances.

File: gui/scripts/syncthing/core/controllers/syncthingController.js [hunk 4]
```diff
@@ -106,6 +105,30 @@ angular.module('syncthing.core')
...
+            if (!restarting) {
+                if (arg.status === 0) {
+                    // A network error, not an HTTP error
+                    $scope.$emit('UIOffline');
+                } else if (arg.status >= 400 && arg.status <= 599) {
+                    // A genuine HTTP error
+                    $('#networkError').modal('hide');
+                    $('#restarting').modal('hide');
+                    $('#shutdown').modal('hide');
```
Comment (AudriusButkevicius) [inline gui/scripts/syncthing/core/controllers/syncthingController.js]: There is more than that, for example settings modal, et al.
AFAIK there is a close all modals jQuery thing.

File: gui/scripts/syncthing/core/controllers/syncthingController.js (hunk 3 above)
Comment (AudriusButkevicius) [inline gui/scripts/syncthing/core/controllers/syncthingController.js]: :+1:

File: gui/scripts/syncthing/core/controllers/syncthingController.js (hunk 4 above)
Comment (AudriusButkevicius) [inline gui/scripts/syncthing/core/controllers/syncthingController.js]: Oh, your old comment got hidden.
I guess we should either hide all, or none (since the stacking works), that's my point.
